### DISCUSSION TITLE
Gets `java.net.ConnectException: Connection refused` on Linux/Mac but works on Windows when migrate to Netty 4.1

### DISCUSSION THREAD

**Discussion by He-Pin** [netty/netty#13526] [Q&A] [answered]
I just send a PR to migrate the pekko/akka `multi-node-testkit` from netty3 to netty4, I think that would be very trival. But it only works on `Windows` and fails 100% on Linux and Mac with `java.net.ConnectException: Connection refused`

The related PR: https://github.com/apache/incubator-pekko/pull/486

`sbt cluster/MultiJvm/test`

https://github.com/he-pin/incubator-pekko/tree/netty4

Tried with different resolver ,not help too.

I tried both Nio transport and native based transport, both will not work on linux/mac. and tried to increase the backlog too very large number too.

Which gives :
```
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3] Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:4711
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3] Caused by: java.net.ConnectException: Connection refused
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:777)
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:337)
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334)
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:776)
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
[info] [JVM-3-NodeLeavingAndExitingMultiJvmNode3]       at java.base/java.lang.Thread.run(Thread.java:829)
```

Update: have been sovled with a `TimeUnit.SECONDS.sleep(2)`, thanks community.

**Comment by hyperxpro** [comment]
Server at `127.0.0.1:4711` TCP rejected the connection by Netty. It would help if you debugged at the network level why the server is not accepting the connection. This is not really Netty-related but networking in general.

**Reply by He-Pin** [reply]
Thanks for chips in, but when I run it against the current `main` branch,which is using netty 3.10.6. the tests passed as expected.

So the I think that may be related  to the `resolver `, tried some, but no progress.

**Reply by hyperxpro** [reply]
Sounds strange. Are you resolving a hostname? What are the answers with Netty 3 and 4 resolvers? Can you tell both IPs?

**Reply by He-Pin** [reply]
> Sounds strange. Are you resolving a hostname? What are the answers with Netty 3 and 4 resolvers? Can you tell both IPs?

Yes , it was resolver a server name,and start a controller as server  and some player as clients.  I thought my port was right but not working on linux and Mac,

I tried netty 4.1.0 and netty 4.1.96 and netty epoll transport ,and customer resolver, all those not works.
 you can checkout my branch and test it.

If you give some input on my line by line migration code, which would be very appreciated. 

I will add some log for the resolved address on both main and the netty4 branch.

I think that may be caused by the resolver not reutning consistence address？

**Reply by hyperxpro** [reply]
Yes, maybe due to some reason, Netty 3 is returning a different address than Netty 4.

Your repo is too big for me to scan. Just log both resolvers addresses and see where things are going wrong.

**Reply by He-Pin** [reply]
> Yes, maybe due to some reason, Netty 3 is returning a different address than Netty 4.
> 
> Your repo is too big for me to scan. Just log both resolvers addresses and see where things are going wrong.

The linked PR is small diff.  Yes the test runs so slow and took nearly 50 minutes  on my windows laptop, and I don't  have a linux or Mac box, so I had to use gitpod for testing, that's why I just made so little progress. 🥲🥲🥲

**Comment by chrisvest** [comment, marked as answer]
The bind() and connect() calls return ChannelFutures which allow you to access the channel before the future completes.
Try waiting for those futures to complete, before progressing further in the test.

**Reply by chrisvest** [reply]
In particular, you probably want for the bind() on the server to complete before calling connect() on the client.

**Reply by He-Pin** [reply]
Thanks and yes, that why there are Await.result in the code :

```scala
  protected def attachConductor(tc: TestConductorExt): Unit = {
    val timeout = tc.Settings.BarrierTimeout.duration
    val startFuture =
      if (selfIndex == 0) tc.startController(initialParticipants, myself, controllerAddr)
      else tc.startClient(myself, controllerAddr)
    try Await.result(startFuture, timeout) //start one by one and wait it complete
    catch {
      case NonFatal(x) => throw new RuntimeException("failure while attaching new conductor", x)
    }
    testConductor = tc
  }
```
But the progress is triggered by `MultiJvmPlugin#multi` and have no time waiting. I will try to add some sleep time there.

**Comment by He-Pin** [comment]
Hi, thank you very much both @hyperxpro and @chrisvest .  the problem is solved after I add a `TimeUnit.SECONDS.sleep(2)`.

The root cause is the multi jvm is started by the `MultiJvmPlugin` in different processes, and seem on windows, it and always ok, but not on Linux/Mac.

After I added a code as below, the problem is gong.
```scala
        if(index == 0) {
          log.info("%s for %s 's started as `Controller`, waiting 2 seconds before start clients.".format(jvmName, testClass))
          TimeUnit.SECONDS.sleep(2)
        }
```

I was think  resolver related thing or my porting is not right, but yes, seems Netty 4 need more time to start the server.

**Reply by He-Pin** [reply]
I have resent the pr: https://github.com/apache/incubator-pekko/pull/539
